### ISSUE TITLE
New semantic analyzer: Support idem forward aliases

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#6404]
This is a follow-up for #6390

This tricky corner case is used in some real code and is supported by old semantic analyzer
```python
C = C
class C:  # type: ignore
    pass

x: C
```
But with new analyzer this results in `Invalid type` since the first definition always wins. Probably we just need to special-case this somehow.

**Comment by ilevkivskyi** [comment]
Another pattern that we need to support with new semantic analyzer is this:
```python
class In:
    ...
class Out:
    In = In  # recognized as variable
```
this currently (both before and after assignment refactor) causes a bunch of troubles in internal code bases. Raising priority to high.

**Comment by ilevkivskyi** [comment]
Oh, wait, this is actually correct, by the rules non-indexed assignments at class scope are always variables, so this was previously a false negative, OK back to normal priority.

**Comment by ilevkivskyi** [comment]
Another repro for this issue with a weird error:
```
x: C
class C:
    pass
C = C
```
fails with
```
main:2: error: Name 'C' already defined on line 4
```
cc @JukkaL

**Comment by JukkaL** [comment]
The examples look meaningless, since the code won't work at runtime. I wonder if the actual code where this comes up is different (involving try/except perhaps). This could also affect the way we'd fix this.

**Comment by ilevkivskyi** [comment]
Yes, the actual code is like
```python
try:
    C = C
except NameError:
    class C:
        ...
```
Also the second [example](https://github.com/python/mypy/issues/6404#issuecomment-500008713) works without try/except.

I think we should fix this before the release since this is actually used in some projects.

**Comment by JukkaL** [comment]
Yes, it would be good to make this better. I'm increasing priority.

But even that example seems incomplete, as that will always generate a `NameError`. I assume it's more like this:

```py
try:
    ModuleNotFoundError = ModuleNotFoundError
except NameError:
    class ModuleNotFoundError(Exception):
        ...
```

The assignment will fail on some Python versions and pass on some.

**Comment by ilevkivskyi** [comment]
Most likely, I don't remember TBH, also `aiohttp` (who first reported this) used to have tons of `*` imports. This may be some other name they define in another module depending on `sys.platform` and then star import it.

**Comment by ilevkivskyi** [comment]
I did some digging and it looks like this happens sometimes after star imports from C acceleration modules.
